diff --git a/simplesamlphp_auth.module b/simplesamlphp_auth.module
index 847c8f5..bcc8c91 100644
--- a/simplesamlphp_auth.module
+++ b/simplesamlphp_auth.module
@@ -149,36 +149,43 @@ function _simplesaml_auth_autoload() {
  * If the user has logged out via SAML, log out the user in Drupal as well.
  */
 function simplesamlphp_auth_init() {
-  if (user_is_anonymous()) {
-    return FALSE;
-  }
 
-  if (!_simplesaml_auth_autoload()) {
+  // Dump all anonymous users out as checking for SAML would waste resources.
+  if (user_is_anonymous()) {
     return FALSE;
   }
 
+  // Intialize user object.
   global $user;
-  global $_simplesamlphp_auth_as;
-  global $_simplesamlphp_auth_saml_attributes;
 
-  // Get users that are allowed default login.
-  $allowed_default_login_users = variable_get('simplesamlphp_auth_allowdefaultloginusers', '');
-  $allowed_uids = explode(",", $allowed_default_login_users);
+  // Check time since user last passed SAML auth (IE: Login). 6 hr default.
+  if($user->login <= (time() - variable_get('simplesamlphp_auth_session_expire', ((60 * 60) * 6)))){
+    if (!_simplesaml_auth_autoload()) {
+      return FALSE;
+    }
 
-  // Get roles that are allowed default login.
-  $allowed_default_login_roles = variable_get('simplesamlphp_auth_allowdefaultloginroles', array());
+    global $_simplesamlphp_auth_as;
+    global $_simplesamlphp_auth_saml_attributes;
 
-  // Check if user is allowed default login.
-  $user_allowed_default_login_roles = array_intersect_key($user->roles, $allowed_default_login_roles);
-  if (variable_get('simplesamlphp_auth_allowdefaultlogin', TRUE) &&
-    (in_array($user->uid, $allowed_uids) || !empty($user_allowed_default_login_roles))) {
-    $user_allowed_default_login = TRUE;
-  }
+    // Get users that are allowed default login.
+    $allowed_default_login_users = variable_get('simplesamlphp_auth_allowdefaultloginusers', '');
+    $allowed_uids = explode(",", $allowed_default_login_users);
+
+    // Get roles that are allowed default login.
+    $allowed_default_login_roles = variable_get('simplesamlphp_auth_allowdefaultloginroles', array());
 
-  // Logs out user if not SAML authenticated and not allowed default login.
-  if (!$_simplesamlphp_auth_as->isAuthenticated() && !$user_allowed_default_login) {
-    module_load_include('pages.inc', 'user');
-    user_logout();
+    // Check if user is allowed default login.
+    $user_allowed_default_login_roles = array_intersect_key($user->roles, $allowed_default_login_roles);
+    if (variable_get('simplesamlphp_auth_allowdefaultlogin', TRUE) &&
+      (in_array($user->uid, $allowed_uids) || !empty($user_allowed_default_login_roles))) {
+      $user_allowed_default_login = TRUE;
+    }
+
+    // Ping SimpleSaml for authentication on user.
+    if (!$_simplesamlphp_auth_as->isAuthenticated() && !isset($user_allowed_default_login)) {
+      module_load_include('pages.inc', 'user');
+      user_logout();
+    }
   }
 }
 
